stubs/psycopg2/psycopg2/extras.pyi

```diff
@@ -28,22 +29,42 @@ from psycopg2._range import (
 
 from .extensions import connection as _connection, cursor as _cursor, quote_ident as quote_ident
 
+_T_cur = TypeVar("_T_cur", bound=_cursor)
+
 class DictCursorBase(_cursor):
     row_factory: Any
     def __init__(self, *args, **kwargs) -> None: ...
     def fetchone(self) -> tuple[Any, ...] | None: ...
     def fetchmany(self, size: int | None = ...) -> list[tuple[Any, ...]]: ...
     def fetchall(self) -> list[tuple[Any, ...]]: ...
-    def __iter__(self): ...
+    def __next__(self) -> tuple[Any, ...]: ...
 
 class DictConnection(_connection):
-    def cursor(self, *args, **kwargs): ...
+    @overload
+    def cursor(self, name: str | bytes | None = ..., *, withhold: bool = ..., scrollable: bool | None = ...) -> DictCursor: ...
+    @overload
+    def cursor(
+        self,
+        name: str | bytes | None = ...,
+        *,
+        cursor_factory: Callable[..., _T_cur],
+        withhold: bool = ...,
+        scrollable: bool | None = ...,
+    ) -> _T_cur: ...
+    @overload
+    def cursor(
+        self, name: str | bytes | None, cursor_factory: Callable[..., _T_cur], withhold: bool = ..., scrollable: bool | None = ...
+    ) -> _T_cur: ...
 
 class DictCursor(DictCursorBase):
     def __init__(self, *args, **kwargs) -> None: ...
     index: Any
     def execute(self, query, vars: Any | None = ...): ...
     def callproc(self, procname, vars: Any | None = ...): ...
+    def fetchone(self) -> DictRow | None: ...  # type: ignore[override]
```

Comment:
An alternative could be to make the base class generic. But that may cause problems until we have TypeVar defaults.